etna/whatson/models.py

```diff
@@ -0,0 +1,621 @@
+    """
+    This model is used to add venue types to event pages.
+    """
+
+    ONLINE = "online", _("Online")
+    IN_PERSON = "in_person", _("In person")
+    HYBRID = "hybrid", _("Hybrid")
+
+
+@register_snippet
+class EventType(models.Model):
+    """
+    This snippet model is used so that editors can add event types,
+    which we use via the event_type ForeignKey to add event types
+    to event pages.
+    """
+
+    name = models.CharField(
+        max_length=255,
+        verbose_name=_("name"),
+    )
+
+    slug = models.SlugField(
+        max_length=255,
+        verbose_name=_("slug"),
+        unique=True,
+    )
+
+    class Meta:
+        verbose_name = _("event type")
+        verbose_name_plural = _("event types")
+
+    def __str__(self):
+        return self.name
+
+
+@register_snippet
+class AudienceType(models.Model):
+    """
+    This snippet model is used so that editors can add audience types,
+    which we use via the audience_type ForeignKey to add audience types
+    to event pages.
+    """
+
+    name = models.CharField(
+        max_length=255,
+        verbose_name=_("name"),
+    )
+
+    slug = models.SlugField(
+        max_length=255,
+        verbose_name=_("slug"),
+        unique=True,
+    )
+
+    class Meta:
+        verbose_name = _("Audience type")
+        verbose_name_plural = _("Audience types")
+
+    def __str__(self):
+        return self.name
+
+
+class AudienceTypeOrderable(Orderable):
+    """
+    This model is used to add multiple audience types to event pages.
+    """
+
+    page = ParentalKey(
+        "wagtailcore.Page",
+        on_delete=models.CASCADE,
+        related_name="event_audience_types",
+    )
+
+    audience_type = models.ForeignKey(
+        "whatson.AudienceType",
+        on_delete=models.CASCADE,
+        related_name="audience_types",
+    )
+
+
+@register_snippet
+class AccessType(models.Model):
+    """
+    This snippet model is used so that editors can add access types,
+    which we use via the AccessTypeOrderable to add multiple access
+    types to event pages.
+    """
+
+    name = models.CharField(
+        max_length=255,
+        verbose_name=_("name"),
+    )
+
+    slug = models.SlugField(
+        max_length=255,
+        verbose_name=_("slug"),
+        unique=True,
+    )
+
+    class Meta:
+        verbose_name = _("Access type")
+        verbose_name_plural = _("Access types")
+
+    panels = [
+        FieldPanel("name"),
+    ]
+
+    def __str__(self):
+        return self.name
+
+
+class AccessTypeOrderable(Orderable):
+    """
+    This model is used to add multiple access types to event pages.
+    """
+
+    page = ParentalKey(
+        "wagtailcore.Page",
+        on_delete=models.CASCADE,
+        related_name="event_access_types",
+    )
+
+    access_type = models.ForeignKey(
+        "whatson.AccessType",
+        on_delete=models.CASCADE,
+        related_name="access_types",
+    )
+
+
+class Host(Orderable):
```

Comment:
`EventHost` most make for a more sensible name here, as they are tied to event pages

Reply:
I'm not entirely sure but I think there might be an ExhibitionPage to come soon, because I was thinking we could have the relation be 
`page = ParentalKey(
        "whatson.EventPage",
        on_delete=models.CASCADE,
        related_name="hosts",
    )`

instead of `wagtailcore.Page`. Is there any benefit of that though? And if there was an ExhibitonPage, would that then need to go back to `wagtailcore.Page`?

Reply:
`wagtailcore.Page` keeps things flexible (you can use the orderable for any page type just by adding `InlinePanel("hosts")` to the panels. Another benefit is that you can prefetch the data even if you are just working with `Page` objects (you don't need specific objects), e.g.:

`events = Page.objects.type(EventPage, ExhibitionPage).prefetch_related("hosts", "speakers")`

Reply:
Coolio - that's what I thought. I shall leave it as is! 😃